src/lang-promql/parser/parser.ts

```diff
@@ -209,14 +209,39 @@ export class Parser {
         this.addDiagnostic(node, 'comparisons between scalars must use BOOL modifier');
       }
     }
-    // TODO missing check regarding cardManyToOne or cardOneToMany
-    // TODO missing check regarding the matching label
+
+    const vectorMatching = buildVectorMatching(this.state, node);
+    if (vectorMatching !== null && vectorMatching.on) {
+      for (const l1 of vectorMatching.matchingLabels) {
+        for (const l2 of vectorMatching.include) {
+          if (l1 === l2) {
+            this.addDiagnostic(node, `label "${l1}" must not occur in ON and GROUP clause at once`);
+          }
+        }
+      }
+    }
+
     if (lt !== ValueType.scalar && lt !== ValueType.vector) {
       this.addDiagnostic(lExpr, 'binary expression must contain only scalar and instant vector types');
     }
     if (rt !== ValueType.scalar && rt !== ValueType.vector) {
       this.addDiagnostic(rExpr, 'binary expression must contain only scalar and instant vector types');
     }
+
+    if ((lt !== ValueType.vector || rt !== ValueType.vector) && vectorMatching !== null) {
+      if (vectorMatching.matchingLabels.length > 0) {
+        this.addDiagnostic(node, 'vector matching only allowed between instant vectors');
+      }
+    } else {
+      if (isSetOperator) {
+        if (vectorMatching?.card === VectorMatchCardinality.CardOneToMany || vectorMatching?.card === VectorMatchCardinality.CardManyToOne) {
+          this.addDiagnostic(node, 'no grouping allowed for this kind of operation');
```

Comment:
I think we could be even clearer here (and probably also improve those messages in Prometheus / LSP):

```suggestion
          this.addDiagnostic(node, 'no grouping allowed for set operations');
```

Reply:
In prometheus it's a bit more clearer since it has in the message the name of the set operator.
But maybe you are thinking about something else ?

Reply:
fixed, thanks for the review :)

Reply:
Ah yeah, that would also be a good option. Either including the name of the operator or just saying it's a set operator would be clearer than just "this kind of operation".

Reply:
Currently I don't have the name of the operator, I just know it's a setOperator. I would suggest to do it a bit later once CMN improves the way to iterate other the tree.

I think there is a lot of refactoring to improve the code in the linter, to make it clearer and simpler once the iteration of the tree is fixed